envisage/tests/test_extension_point.py

```diff
@@ -253,6 +328,80 @@ class Foo(TestBase):
 
         self.assertEqual([42], registry.get_extensions("my.ep"))
 
+    def test_set_typed_extension_point_emit_change(self):
+        """ Test change event is emitted for setting the extension point """
+
+        registry = self.registry
+
+        # Add an extension point.
+        registry.add_extension_point(self._create_extension_point("my.ep"))
+
+        # Declare a class that consumes the extension.
+        class Foo(TestBase):
+            x = ExtensionPoint(List(Int), id="my.ep")
+
+        on_trait_change_events = []
+
+        def on_trait_change_handler(*args):
+            on_trait_change_events.append(args)
+
+        observed_events = []
+
+        f = Foo()
+        f.on_trait_change(on_trait_change_handler, "x")
+        f.observe(observed_events.append, "x")
+
+        # when
+        ExtensionPoint.connect_extension_point_traits(f)
+
+        # then
+        self.assertEqual(len(on_trait_change_events), 1)
+        self.assertEqual(len(observed_events), 1)
+        event, = observed_events
+        self.assertEqual(event.object, f)
+        self.assertEqual(event.name, "x")
+        self.assertEqual(event.old, Undefined)
+        self.assertEqual(event.new, [])
+
+    def test_object_garbage_collectable(self):
+        """ object can be garbage collected after disconnecting listeners."""
+        registry = self.registry
+
+        # Add an extension point.
+        registry.add_extension_point(self._create_extension_point("my.ep"))
+
+        # Declare a class that consumes the extension.
+        class Foo(TestBase):
+            x = ExtensionPoint(List(Int), id="my.ep")
+
+        f = Foo()
+        object_ref = weakref.ref(f)
+
+        # when
+        ExtensionPoint.connect_extension_point_traits(f)
+        ExtensionPoint.disconnect_extension_point_traits(f)
+        del f
+
+        # then
+        self.assertIsNone(object_ref())
+
+    def test_object_pickability(self):
+        # Add an extension point.
+        self.registry.add_extension_point(ExtensionPoint(id="my.ep"))
+
+        # An object is created, connected to the registry and have the
+        # extension point created.
+        f = ClassWithExtensionPoint(extension_registry=self.registry)
+        ExtensionPoint.connect_extension_point_traits(f)
+        self.registry.set_extensions("my.ep", [1, 2, 3])
+        self.assertEqual(f.x, [1, 2, 3])
+
+        # then
+        for protocol in range(pickle.HIGHEST_PROTOCOL + 1):
+            serialized = pickle.dumps(f.x, protocol=protocol)
+            deserialized = pickle.loads(serialized)
+            self.assertEqual(deserialized, [1, 2, 3])
```

Comment:
When this test was run on Python 3.6 (the one built from `etstool.py`), the test passes.
When I try to run it on Python 3.8 and Python 3.9 after pip-installing envisage from source, this test fails:
```
======================================================================
FAIL: test_object_pickability (envisage.tests.test_extension_point.ExtensionPointTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kchoi/Work/ETS/envisage/envisage/tests/test_extension_point.py", line 403, in test_object_pickability
    self.assertEqual(deserialized, [1, 2, 3])
AssertionError: [] != [1, 2, 3]
```

Reply:
Same failure occurs to a Python 3.7 venv, when the test is run, this RuntimeWarning appears:
```
test_object_pickability (envisage.tests.test_extension_point.ExtensionPointTestCase) ... /Users/kchoi/Work/ETS/envisage/envisage/tests/test_extension_point.py:402: RuntimeWarning: Extension point cannot be mutated directly.
  deserialized = pickle.loads(serialized)
FAIL
...
======================================================================
FAIL: test_object_pickability (envisage.tests.test_extension_point.ExtensionPointTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kchoi/Work/ETS/py37-venv/envisage-env/lib/python3.7/site-packages/envisage/tests/test_extension_point.py", line 403, in test_object_pickability
    self.assertEqual(deserialized, [1, 2, 3])
AssertionError: [] != [1, 2, 3]
```

Reply:
Fixed.